prusti-encoder/src/encoders/mir_pure.rs

```diff
@@ -315,46 +315,51 @@ impl<'vir, 'enc> Encoder<'vir, 'enc>
         self.reify_binds(res, self.mk_local_ex(mir::RETURN_PLACE, ret_version))
     }
 
-    fn find_join_point(
-        &self,
-        dominators: &Dominators<mir::BasicBlock>,
-        start: mir::BasicBlock,
-        end: mir::BasicBlock,
-    ) -> mir::BasicBlock {
-        dominators.dominators(end)
-            .take_while(|bb| *bb != start)
-            .last()
-            .unwrap()
-    }
-
     fn encode_cfg(
         &mut self,
         curr_ver: &HashMap<mir::Local, usize>,
-        start: mir::BasicBlock,
-        end: mir::BasicBlock,
-    ) -> Update<'vir> {
+        curr: mir::BasicBlock,
+        branch_point: Option<mir::BasicBlock>,
+    ) -> (mir::BasicBlock, Update<'vir>) {
         let dominators = self.body.basic_blocks.dominators();
+        // We should never actually reach the join point bb: we should catch
+        // this case and stop recursion in the `Goto` branch below. If this
+        // assert fails we we may need to add catches in the other branches.
+        debug_assert!(match (dominators.immediate_dominator(curr), branch_point) {
+            (Some(immediate_dominator), Some(branch_point)) if immediate_dominator == branch_point =>
+                // We could also be immediately dominated by the join point if we
+                // are the next bb right after the `SwitchInt`.
+                self.body.basic_blocks.predecessors()[curr].contains(&branch_point),
+            _ => true,
+        }, "reached join point bb {curr:?} (bp {branch_point:?})");
 
         // walk block statements first
         let mut new_curr_ver = curr_ver.clone();
-        let stmt_update = self.body[start].statements.iter()
+        let stmt_update = self.body[curr].statements.iter()
             .fold(Update::new(), |update, stmt| {
                 let newer = self.encode_stmt(&new_curr_ver, stmt);
                 newer.add_to_map(&mut new_curr_ver);
                 update.merge(newer)
             });
 
         // then walk terminator
-        let term = self.body[start].terminator.as_ref().unwrap();
+        let term = self.body[curr].terminator.as_ref().unwrap();
         match &term.kind {
-            mir::TerminatorKind::Goto { target } => {
-                if *target == end {
-                    // We are done with the current fragment of the CFG, the
-                    // rest is handled in a parent call.
-                    return stmt_update;
+            &mir::TerminatorKind::Goto { target } => {
+                match (dominators.immediate_dominator(target), branch_point) {
+                    // As soon as we are about to step to a bb where
+                    (Some(immediate_dominator), Some(branch_point))
+                        if immediate_dominator == branch_point =>
```

Comment:
Can you not simply `branch_point == dominators.immediate_dominator(target)`?

Reply:
No, we do not want `true` if both are `None`